### PR TITLE
info: show correct file origin for formula from path or url

### PR DESCRIPTION
**Pull request by xu-cheng**

### CODE REVIEW

File: Library/Homebrew/cmd/info.rb [hunk 1]
```diff
@@ -62,16 +62,18 @@ def github_fork
   end
 
   def github_info f
-    if f.tap?
-      user, repo = f.tap.split("/", 2)
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/info.rb]: `f.tap?` and `f.tap` would do the same regex matching twice. So squash them into one in the benefit of performance.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/info.rb]: I don't think performance really matters in code like this but given we're changing it anyway it seems fair enough.

File: Library/Homebrew/cmd/info.rb [hunk 2]
```diff
@@ -62,16 +62,18 @@ def github_fork
...
       path = f.path.relative_path_from(HOMEBREW_LIBRARY.join("Taps", f.tap))
-    else
+      "https://github.com/#{user}/#{repo}/blob/master/#{path}"
+    elsif f.core_formula?
       user = f.path.parent.cd { github_fork }
       repo = "homebrew"
       path = f.path.relative_path_from(HOMEBREW_REPOSITORY)
+      "https://github.com/#{user}/#{repo}/blob/master/#{path}"
+    else
+      "file://#{f.path}"
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/info.rb]: Mmm, not sure what I think about a `file://` URL here instead of just a path.

Comment (xu-cheng) [inline Library/Homebrew/cmd/info.rb]: Using `file://`, so `exec_browser(*ARGV.formulae.map { |f| github_info(f) })` still works.

File: Library/Homebrew/cmd/info.rb [hunk 3]
```diff
@@ -62,16 +62,17 @@ def github_fork
...
       path = f.path.relative_path_from(HOMEBREW_LIBRARY.join("Taps", f.tap))
-    else
+      "https://github.com/#{user}/#{repo}/blob/master/#{path}"
+    elsif f.core_formula?
       user = f.path.parent.cd { github_fork }
-      repo = "homebrew"
       path = f.path.relative_path_from(HOMEBREW_REPOSITORY)
+      "https://github.com/#{user}/homebrew/blob/master/#{path}"
+    else
+      "file://#{f.path}"
```
Comment (jacknagel) [inline Library/Homebrew/cmd/info.rb]: I don't think we need `file://` here. Won't that make it more difficult to copy/paste in the terminal?

File: Library/Homebrew/cmd/info.rb (hunk 1 above)
Comment (jacknagel) [inline Library/Homebrew/cmd/info.rb]: The idea here is to minimize the number of places that have to know how to extract the tap user/repo from the formula path, so I'd prefer not to make this change.

File: Library/Homebrew/cmd/info.rb (hunk 3 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/info.rb]: Use `file://` to make `exec_browser(*ARGV.formulae.map { |f| github_info(f) })` still works.

File: Library/Homebrew/cmd/info.rb (hunk 1 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/info.rb]: I just think it's redundancy to repeat the regex matching.

Comment (jacknagel) [inline Library/Homebrew/cmd/info.rb]: Why should downstream code need to know that the tap name is determined by regex matching the path to the formula's location on disk, when it can just ask the formula for the tap name?

File: Library/Homebrew/cmd/info.rb (hunk 3 above)
Comment (jacknagel) [inline Library/Homebrew/cmd/info.rb]: Doesn't it seem weird for `brew info --github` to open a local file in the browser?

File: Library/Homebrew/cmd/info.rb (hunk 1 above)
Comment (mikemcquaid) [inline Library/Homebrew/cmd/info.rb]: I agree. If there was an optimisation here it'd be making it so `f.tap?` and `f.tap` use a shared cache but I think that's overkill on a simple string match.

File: Library/Homebrew/cmd/info.rb (hunk 3 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/info.rb]: Not too weird to me. What's your suggestion?

File: Library/Homebrew/cmd/info.rb (hunk 1 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/info.rb]: I understand you want to public less API as possible. But we are changing the internal code here not downstream.

Comment (jacknagel) [inline Library/Homebrew/cmd/info.rb]: This doesn't have anything to do with "public API". By "downstream" I mean code that is consuming the formula API, internal code included. There is no reason that consuming code needs to know these implementation details.

Please undo this change.

Comment (xu-cheng) [inline Library/Homebrew/cmd/info.rb]: How can consuming code know or not know the implementation details? I still think avoiding duplicated computation is worthy to do.

However, since you are the maintainers. It's your call. I will revert it.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/info.rb]: > However, since you are the maintainer. It's your call. I will revert it.

Thanks!

**Comment by xu-cheng** [comment]
Any comment on this one?

**Comment by mikemcquaid** [comment]
I think let's remove the `file://`

**Comment by xu-cheng** [comment]
> I think let's remove the file://

Should I do any change on `exec_browser(*ARGV.formulae.map { |f| github_info(f) })`

**Comment by mikemcquaid** [comment]
@xu-cheng Yeh, I think perhaps that should just ignore non-http/non-github URLs.

**Comment by xu-cheng** [comment]
Oh, I just tested. `exec_browser` will open the file with or without `file://`. So this PR is ready to be merged.
